[ML] TransformIT » testStopWaitForCheckpoint test fails

**Build scan**:
https://gradle-enterprise.elastic.co/s/btdxhp7hfpb6y
**Repro line**:
```
./gradlew ':x-pack:plugin:transform:qa:multi-node-tests:javaRestTest' --tests "org.elasticsearch.xpack.transform.integration.TransformIT.testStopWaitForCheckpoint" -Dtests.seed=E5346188868E92A8 -Dtests.security.manager=true -Dtests.locale=pt-BR -Dtests.timezone=America/Thule -Druntime.java=8
```
**Reproduces locally?**:
No
**Applicable branches**:
7.x (though probably master as well)
**Failure history**:

**Failure excerpt**:
```
org.elasticsearch.xpack.transform.integration.TransformIT > testStopWaitForCheckpoint FAILED |  
-- | --
  | ElasticsearchStatusException[Elasticsearch exception [type=status_exception, reason=Failed to update transform task [transform-wait-for-checkpoint] state value should_stop_at_checkpoint from [true] to [true]]] |  
  | at __randomizedtesting.SeedInfo.seed([E5346188868E92A8:98E14DE34A1A789E]:0) |  
  | at org.elasticsearch.rest.BytesRestResponse.errorFromXContent(BytesRestResponse.java:187) |  
  | at org.elasticsearch.client.RestHighLevelClient.parseEntity(RestHighLevelClient.java:1911) |  
  | at org.elasticsearch.client.RestHighLevelClient.parseResponseException(RestHighLevelClient.java:1888) |  
  | at org.elasticsearch.client.RestHighLevelClient.internalPerformRequest(RestHighLevelClient.java:1645) |  
  | at org.elasticsearch.client.RestHighLevelClient.performRequest(RestHighLevelClient.java:1617) |  
  | at org.elasticsearch.client.RestHighLevelClient.performRequestAndParseEntity(RestHighLevelClient.java:1584) |  
  | at org.elasticsearch.client.TransformClient.stopTransform(TransformClient.java:319) |  
  | at org.elasticsearch.xpack.transform.integration.TransformIntegTestCase.stopTransform(TransformIntegTestCase.java:136) |  
  | at org.elasticsearch.xpack.transform.integration.TransformIT.testStopWaitForCheckpoint(TransformIT.java:288)
```
This indicates some race condition when trying to set the value for `should_stop_at_checkpoint`. I am not sure if it was retried in the call and then caused the failure or if two callers are making the stop call.

Getting logging to attach to the issue.

The log shows an off by 2, so the doc has been updated twice by another caller:

```
Caused by: org.elasticsearch.index.engine.VersionConflictEngineException: [data_frame_transform_state_and_stats-transform-wait-for-checkpoint]: version conflict, required seqNo [16], primary term [1]. current document has seqNo [18] and primary term [1]
```

The "other caller" can only be the stop transform transport although it _should_ only update the doc directly if there is no indexer. Looks like some race condition that makes the transport think that there is no indexer.

The logs are not very helpful otherwise, I suggest we add some debug logging to e.g. log whenever the state doc has changed, including `seqNo`.

--- a/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java
+++ b/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java
@@ -588,6 +588,22 @@ public abstract class TransformIndexer extends AsyncTwoPhaseIndexer<TransformInd
             return false;
         }
 
+        /*
+         * ignore if indexer thread is shutting down (after finishing a checkpoint)
+         * shutting down means:
+         *  - indexer has finished a checkpoint and called onFinish
+         *  - indexer state has changed from indexing to started
+         *  - state persistence has been called but has _not_ returned yet
+         *
+         *  If we trigger the indexer in this situation the 2nd indexer thread might
+         *  try to save state at the same time, causing a version conflict
+         *  see gh#67121
+         */
+        if (indexerThreadShuttingDown) {
+            logger.debug("[{}] indexer thread is shutting down. Ignoring trigger.", getJobId());
+            return false;
+        }
+
         return super.maybeTriggerAsyncJob(now);
     }
 

